fabtests/unit/ep_test.c

```diff
@@ -0,0 +1,424 @@
+
+#include <rdma/fi_errno.h>
+
+#include "unit_common.h"
+#include "shared.h"
+
+#define EP_TEST_SIZES_STRIDE 101
+
+static unsigned int num_eps = 128;
+static unsigned int max_eps;
+static unsigned int seed = -1;
+
+struct ft_res {
+	struct fid_ep *ep;
+	struct fid_av *av;
+	struct fid_cq *txcq;
+	struct fid_cq *rxcq;
+} *res;
+
+static char err_buf[512];
+
+static int create_av(struct fid_av **av, enum fi_av_type type, int rx_ctx_bits,
+		     size_t count, size_t ep_per_node, const char *name,
+		     void *map_addr, uint64_t flags)
+{
+	struct fi_av_attr av_attr;
+
+	memset(&av_attr, 0, sizeof(av_attr));
+	av_attr.type = type;
+	av_attr.rx_ctx_bits = rx_ctx_bits;
+	av_attr.count = count;
+	av_attr.ep_per_node = ep_per_node;
+	av_attr.name = name;
+	av_attr.map_addr = map_addr;
+	av_attr.flags = flags;
+
+	return fi_av_open(domain, &av_attr, av, NULL);
+}
+
+static int create_cq(struct fid_cq **cq, size_t size, uint64_t flags,
+		     enum fi_cq_format format, enum fi_wait_obj wait_obj)
+{
+	struct fi_cq_attr cq_attr;
+
+	memset(&cq_attr, 0, sizeof(cq_attr));
+	cq_attr.size = size;
+	cq_attr.flags = flags;
+	cq_attr.format = format;
+	cq_attr.wait_obj = wait_obj;
+
+	return fi_cq_open(domain, &cq_attr, cq, NULL);
+}
+
+static void close_ep(int idx)
+{
+	FT_CLOSE_FID(res[idx].ep);
+	FT_CLOSE_FID(res[idx].av);
+	FT_CLOSE_FID(res[idx].txcq);
+	FT_CLOSE_FID(res[idx].rxcq);
+}
+
+static int open_ep(int idx)
+{
+	int ret;
+
+	struct ft_res *tres = &res[idx];
+
+	fi->tx_attr->op_flags = 0;
+	ret = fi_endpoint(domain, fi, &tres->ep, NULL);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "fi_endpoint failed when opening new EP", ret);
+		return ret;
+	}
+
+	ret = create_av(&tres->av, FI_AV_UNSPEC, 0, fi->domain_attr->ep_cnt,
+			fi->domain_attr->ep_cnt, NULL, NULL, 0);
+	if (ret)
+		goto out_close_ep;
+	ret = create_cq(&tres->txcq, fi->tx_attr->size, 0, FI_CQ_FORMAT_UNSPEC, FI_WAIT_NONE);
+	if (ret)
+		goto out_close_ep;
+	ret = create_cq(&tres->rxcq, fi->rx_attr->size, 0, FI_CQ_FORMAT_UNSPEC, FI_WAIT_NONE);
+	if (ret)
+		goto out_close_ep;
+
+	ret = ft_enable_ep(tres->ep, NULL, tres->av, tres->txcq, tres->rxcq, NULL, NULL);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "fi_enable_ep failed when opening new EP", ret);
+		goto out_close_ep;
+	}
+
+	return 0;
+
+out_close_ep:
+	close_ep(idx);
+	return ret;
+}
+
+struct ep_sizes_pair {
+	size_t tx_size;
+	size_t rx_size;
+};
+
+static int ep_open_sizes(void)
+{
+	size_t size, max_size;
+	size_t max_tx_size = fi->tx_attr->size, max_rx_size = fi->rx_attr->size;
+	int testret = FAIL;
+	int i, ret, ep_idx = 0;
+	struct ep_sizes_pair expected_success[] = {
+			/* zero sized QP it not yet supported
+			 * {0, max_rx_size},
+			 * {max_tx_size, 0},
+			 */
+			{max_tx_size, max_rx_size}
+	};
+	struct ep_sizes_pair expected_failure[] = {
+			{max_tx_size + 1, max_rx_size},
+			{max_tx_size, max_rx_size + 1}
+	};
+
+	max_size = MAX(max_tx_size, max_rx_size);
+	for (size = 1; size <= max_size; size += EP_TEST_SIZES_STRIDE) {
```

Comment:
The 'max' sizes can be 64k or higher.  Walking through that many sizes 101 at a time will be time consuming.  We can use the existing test_sizes array here instead.  Because that array consists of values that are (2^n) and (2^n + 2 (n-1)), we can test each value as well as each value with a fixed value added to it in order to stress unlikely values.